### PR TITLE
Fix #682: no fee for frozen account related operations

### LINKED ISSUE #682: Apply no-fee to a transaction related to frozen account.
Apply no-fee to a transaction related to frozen account.
It affects freezing, unfreezing request and unfreezing payment.

### PR DESCRIPTION
**Pull request by kfangw**
### Github Issue
#682 

### Background
* No fee required for `freezing` `unfreezing` operation.

### CODE REVIEW

**Comment by kfangw** [comment]
@spikeekips @soonkuk @anarcher 
Please review this PR as soon as possible  ^O^

@outsideris is waiting for this

File: lib/node/runner/checker_ballot_transaction.go [hunk 1]
```diff
@@ -265,37 +277,23 @@ func ValidateTx(st *storage.LevelDBBackend, tx transaction.Transaction) (err err
...
+
 	switch op.H.Type {
 	case operation.TypeCreateAccount:
 		var ok bool
 		var casted operation.CreateAccount
 		if casted, ok = op.B.(operation.CreateAccount); !ok {
 			return errors.TypeOperationBodyNotMatched
 		}
+		if source.Linked != "" {
+			return errors.ForozenAccountCannotCreateAccount
```
Comment (anarcher) [inline lib/node/runner/checker_ballot_transaction.go]:
```suggestion
			return errors.FrozenAccountCannotCreateAccount
```

File: lib/node/runner/checker_ballot_transaction.go [hunk 2]
```diff
@@ -310,16 +308,16 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op opera
...
-			var expected common.Amount
-			expected, err = source.Balance.Sub(common.BaseFee)
-			if casted.Amount != expected {
+			if source.Balance != casted.Amount { // No fee for Unfreeze Payment
 				return errors.FrozenAccountMustWithdrawEverything
 			}
 			// Unfreezing must be done after X period from unfreezing request
 			iterFunc, closeFunc := block.GetBlockOperationsBySource(st, source.Address, nil)
-			bo, _, _ := iterFunc()
+			bo, _, _ := iterFunc() //Get the first operation submitted by the source(frozen) account
```
Comment (anarcher) [inline lib/node/runner/checker_ballot_transaction.go]: It is possible that `bo is nil`?

Comment (kfangw) [reply lib/node/runner/checker_ballot_transaction.go]: https://github.com/bosnet/sebak/blob/13dbe0a17626858b51b61b4dde078554b5597458/lib/block/operation.go#L229-233
It always returns not nil.
If there is no item then BlockOperation{} is returned

**Comment by soonkuk** [comment]
@kfangw I do following test in standalone mode.

1-1. genesis account create general account1 with amount 10000
1-2. general account1 create frozen account with amount 10000
1-3. frozen account request unfreezing
1-4. frozen account refund to general account1

It works fine.

2-1. genesis account create general account1 with amount 10000
2-2. general account1 create frozen account with amount 10000
2-3. frozen account request unfreezing
2-4 genesis account create general account2
2-4. frozen account refund to general account2

It works fine.

3-1. genesis account create general account1 with amount 10000
3-2. general account1 create frozen account with amount 10000
3-3. frozen account request unfreezing
3-4. frozen account try to refund to general account2
3-4 genesis account create general account2
3-5. frozen account try to refund to general account2

Then I get the error 
```2018/11/17 02:08:30 Network error: {"code":163,"message":"http failed to get response","data":{"status":400}} body: {"type":"https://boscoin.io/sebak/error/106","title":"received new, but known message","status":400}```

**Comment by outsideris** [comment]
Currently, I tested freezing and unfreezing with zero fee. It works. 
I don't test withdraw all from frozen account yet.

**Comment by kfangw** [comment]
I will add integration test case including what soonkuk mentioned above.

**Comment by outsideris** [comment]
I tested withdrawing from unfrozen accounts. It works in a client side. Thank you.

**Comment by kfangw** [comment]
> @kfangw I guess the reason of above error is a `TransactionHistory`.
> 
> [sebak/lib/node/runner/api_node.go](https://github.com/bosnet/sebak/blob/13dbe0a17626858b51b61b4dde078554b5597458/lib/node/runner/api_node.go#L134)
> 
> Line 134 in [13dbe0a](/bosnet/sebak/commit/13dbe0a17626858b51b61b4dde078554b5597458)
> 
>  block.SaveTransactionHistory(api.storage, checker.Transaction, block.TransactionHistoryStatusRejected) 
> I think this saves a `TransactionHistory` of the rejected status, when error occurs.
> But from a frozen account, the exact same payment transaction (with same sequence id) for unfreezing must be generated after the previous payment transaction is rejected.
> And because the previous transaction was saved rejected, the next same transaction can not be submitted.

please refer this #754

**Comment by outsideris** [comment]
Please ping me if you done this job. I will test it again in client.

**Comment by kfangw** [comment]
@anarcher @soonkuk I have done with this PR.
I changed it as `SIMPLE` as possible.

- Freezing has no fee. It is only no fee operation ever!
- The frozen account cannot create or pay to another account
- The only original account can receive the whole amount of frozen balance

-- Removed `context` changes. It is not related to this PR. hence, I will make another PR for it. Only freezing(create) operation has no fee, hence, fee validation can be done at IsWellFormed checker @anarcher @soonkuk

File: cmd/sebak/cmd/wallet/unfreezeRequest.go [hunk 3]
```diff
@@ -130,7 +130,7 @@ func makeTransactionUnfreezingRequest(kpSource keypair.KP, seqid uint64) transac
 
 	txBody := transaction.Body{
 		Source:     kpSource.Address(),
-		Fee:        common.Amount(common.BaseFee),
+		Fee:        common.FrozenFee,
```
Comment (soonkuk) [inline cmd/sebak/cmd/wallet/unfreezeRequest.go]: This should be changed to `common.Amount(common.BaseFee)`

Comment (kfangw) [reply cmd/sebak/cmd/wallet/unfreezeRequest.go]: done

File: lib/block/account.go [hunk 4]
```diff
@@ -43,6 +43,10 @@ func NewBlockAccountLinked(address string, balance common.Amount, linked string)
 	}
 }
 
+func (b *BlockAccount) IsFrozen() bool {
```
Comment (anarcher) [inline lib/block/account.go]: :+1:

File: cmd/sebak/cmd/wallet/payment.go [hunk 5]
```diff
@@ -170,8 +169,16 @@ func init() {
...
+func makeTransactionCreateAccount(kpSource keypair.KP, kpDest keypair.KP, amount common.Amount, seqid uint64, isFrozen bool) transaction.Transaction {
+	var opb operation.CreateAccount
+	var fee common.Amount
+	if isFrozen {
+		opb = operation.NewCreateAccount(kpDest.Address(), amount, kpSource.Address())
+		fee = common.FrozenFee
+	} else {
+		opb = operation.NewCreateAccount(kpDest.Address(), amount, "")
+		fee = common.BaseFee
+	}
```
Comment (Geod24) [inline cmd/sebak/cmd/wallet/payment.go]: That's not an improvement. What's wrong with the previous version ?

File: lib/errors/errors.go [hunk 6]
```diff
@@ -102,4 +102,8 @@ var (
 	InflationPFAmountMissMatched              = NewError(184, "inflation pf amount missmatched")
 	InflationPFFundingAddressMissMatched      = NewError(185, "inflation pf funding address missmatched")
 	CongressAddressMisMatched                 = NewError(186, "congress address mismatched")
+	FrozenAccountHasMoreThanOneOperation      = NewError(187, "Transaction from frozen account can have only one operation")
+	FrozenAccountCannotCreateAccount          = NewError(188, "Frozen account cannot create another account")
+	InflationPFOperationNotPermitted          = NewError(189, "Inflation PF operation is not permitted")
+	FrozenAccountMustWithdrawToOriginal       = NewError(190, "frozen account can only withdraw to the original account")
```
Comment (Geod24) [inline lib/errors/errors.go]: We discussed that before. Why do you want to change it now ?

File: lib/node/runner/checker_ballot_transaction.go [hunk 7]
```diff
@@ -267,35 +274,24 @@ func ValidateTx(st *storage.LevelDBBackend, config common.Config, tx transaction
...
 	switch op.H.Type {
 	case operation.TypeCreateAccount:
 		var ok bool
 		var casted operation.CreateAccount
 		if casted, ok = op.B.(operation.CreateAccount); !ok {
 			return errors.TypeOperationBodyNotMatched
 		}
-		if exists, err := block.ExistsBlockAccount(st, op.B.(operation.CreateAccount).Target); err == nil && exists {
+
+		var exists bool
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: That change is unrelated

File: lib/node/runner/checker_ballot_transaction.go [hunk 8]
```diff
@@ -267,35 +274,24 @@ func ValidateTx(st *storage.LevelDBBackend, config common.Config, tx transaction
...
-			if lastblock.Height-bo.Height < common.UnfreezingPeriod {
-				return errors.UnfreezingNotReachedExpiration
-			}
-			// If it's a frozen account we check that only whole units are frozen
-			if casted.Linked != "" && (casted.Amount%common.Unit) != 0 {
-				return errors.FrozenAccountCreationWholeUnit // FIXME
-			}
+
+		if source.IsFrozen() {
+			return errors.FrozenAccountCannotCreateAccount
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: Why not ?

File: lib/node/runner/checker_ballot_transaction.go [hunk 9]
```diff
@@ -213,6 +213,13 @@ func ValidateTx(st *storage.LevelDBBackend, config common.Config, tx transaction
 		return errors.BlockAccountDoesNotExists
 	}
 
+	// check, multiple operation from frozen account.
+	if ba.IsFrozen() {
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: Why is this checked ? Was there a bug before ?

File: cmd/sebak/cmd/wallet/payment.go (hunk 5 above)
Comment (kfangw) [reply cmd/sebak/cmd/wallet/payment.go]: IMO, To calculate fee, it is better to read

File: lib/errors/errors.go (hunk 6 above)
Comment (kfangw) [reply lib/errors/errors.go]: To make it simple.
When we let frozen account do functions as ordinal account then,
we should think about all of the side effects.

File: lib/node/runner/checker_ballot_transaction.go (hunk 9 above)
Comment (kfangw) [reply lib/node/runner/checker_ballot_transaction.go]: Originally it was @soonkuk's change. 
It can be removed i think

File: lib/node/runner/checker_ballot_transaction.go (hunk 7 above)
Comment (kfangw) [reply lib/node/runner/checker_ballot_transaction.go]: agreed.

File: lib/node/runner/checker_ballot_transaction.go (hunk 8 above)
Comment (kfangw) [reply lib/node/runner/checker_ballot_transaction.go]: um... it can. agreed.

File: lib/errors/errors.go (hunk 6 above)
Comment (Geod24) [reply lib/errors/errors.go]: This has already been done. And it's not making it simpler, you are adding more code to support that restriction.

**Comment by Geod24** [comment]
@kfangw : Yeah I have no problem with change `1`.But `2` and `3` don't seem related to marketing, did they ask for it as well ?

**Comment by kfangw** [comment]
> @kfangw : Yeah I have no problem with change `1`.But `2` and `3` don't seem related to marketing, did they ask for it as well ?

only 1 is required.
2, 3 was aroze when we think about no fee for `unfreezing request` and `refund(payment)` operation.
I discussed with other team and decided that only freezing operation has no fee.
So, 2, 3 are no longer needed.
@Geod24 

please review PR with the lastest commit

File: cmd/sebak/cmd/wallet/payment.go (hunk 5 above)
Comment (Geod24) [reply cmd/sebak/cmd/wallet/payment.go]: Ok 👍

File: tests/client/self.env [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+# This is expanded by the `entrypoint.sh` script based on spaces
+SEBAK_VALIDATORS=self
+# This node's public key
+SEBAK_GENESIS_BLOCK=GDIRF4UWPACXPPI4GW7CMTACTCNDIKJEHZK44RITZB4TD3YUM6CCVNGJ
+# common account
+SEBAK_COMMON_ACCOUNT=GCTVCU764UPXKRJW5DS5AWF5ETCCIYPZTWXN5U5CLUNAAJH6D5NGEIIH
+# congress account
+SEBAK_CONGRESS_ADDR=GAIJAH4FCEB3AWS2NVKRNDGVJ5QPD7VNOQNGZEPL2FAI6GPZQNKEXRWI
+# unfreezing period
+SEBAK_UNFREEZING_PERIOD=0
```
[the remark this answers is not in the document]
Comment (kfangw) [reply tests/client/self.env]: It is intended for the test.

[the remark this answers is not in the document]
Comment (Geod24) [reply tests/client/self.env]: But that does not mirror any possible usage, and 0 is quite a special value. Make it 1 or 2, at least.

**Comment by kfangw** [comment]
When @Geod24 approved then, I will merge this PR.
@outsideris FYI.

**Comment by outsideris** [comment]
I re-add fees for unfreezing and withdraw. Unfreezing is fine and I'm waiting unfreezing period for withdrawing.

**Comment by outsideris** [comment]
Withdraw is works. 

However, I found some mismatch. So, I want to remove a fee from unfreezing.(Freezing and unfreezing has no-fee, and payment has fee.)

* If we use fee for unfreezing, the balance is reduce to 9999.999 BOS from 10,000 BOS. 
* But last 2 weeks is included in reward period. So, it make sto hard to calculate how much reward.
* The pair for freezing and unfreezing doesn't have a fee is make sense and simple.

**Comment by Geod24** [comment]
@outsideris : I don't understand your message. You want to remove fee from unfreezing, yet mention "Freezing and unfreezing has no-fee, and payment has fee.". So I guess you meant "Freezing and unfreezing request has no-fee, ..." ?

In this case, do you want to remove the fee from the payment ?

> But last 2 weeks is included in reward period. So, it make sto hard to calculate how much reward.

I don't understand the connection here. Rewards go to the linked account.

**Comment by outsideris** [comment]
Regardless we still discuss about removing a fee for unfreezing, this PR works for freezing/unfreezing without fee and payment with fee.
